algebra2/gaussian_elimination/gaussian_elimination.h

```diff
@@ -0,0 +1,566 @@
+auto determinant(ranges::matrix_view<ranges::fraction<T>, LP> m) -> std::expected<ranges::fraction<T>, algorithms::gaussian_elimination::error>{
+    ranges::fraction<T> det{1, 1}, res;
+
+    std::println("//////////Determinant of a matrix m containing fractions//////////\n");
+    std::println("Matrix m = {}\n", m);
+
+    if(m.extent(0) == m.extent(1)){
+        for(int i = 0;  i < m.extent(0); i++){
+            if(m[i][i].numerator == 0){
+                for(int j = i+1; j < m.extent(0); j++){
+                    if(m[j][i].numerator != 0){
+                        swap(m, i, j);
+                        break;
+                    }
+                }
+            }
+            if(m[i][i].numerator != 0){
+                for(int k = i+1; k < m.extent(0); k++){
+                    res = m[k][i]/m[i][i];
+                    if(res.numerator != 0){
+                        subtract(m, k, i, res);
+                    }
+                }
+            }
+        }
+
+        for(int i = 0; i < m.extent(0); i++){
+            det *= m[i, i];
+        }
+        std::println("The determinant of m = {}\n", det);
+        return det;
+    }else{
+        std::println("Cannot calculate a determinant of a non square matrix\n");
+        return std::unexpected(algorithms::gaussian_elimination::error::not_square);
+    }
+}
+/*
+ * description:
+ *      Calculates determinant of matrix m*/
+template <typename T, typename LP>
+auto determinant(ranges::matrix_view<T, LP> m) -> std::expected<T, algorithms::gaussian_elimination::error>{
+    double det = 1, res;
+
+    std::println("//////////Determinant of matrix m//////////\n");
+    std::println("Matrix m = {}\n", m);
+
+    if(m.extent(0) == m.extent(1)){
+
+        for(int i = 0;  i < m.extent(0); i++){
+            if(m[i, i] == 0){
+                for(int j = i+1; j < m.extent(0); j++){
+                    if(m[j, i] != 0){
+                        swap(m, i, j);
+                        break;
+                    }
+                }
+            }
+            if(m[i, i] != 0){
+                for(int k = i+1; k < m.extent(0); k++){
+                    res = m[k, i]/m[i, i];
+                    if(res != 0){
+                        subtract(m, k, i, res);
+                    }
+                }
+            }
+        }
+
+        for(int i = 0; i < m.extent(0); i++){
+            det *= m[i, i];
+        }
+        std::println("The determinant of m = {}\n", det);
+        return det;
+    }else{
+        std::println("Cannot calculate a determinant of a non square matrix\n");
+        return std::unexpected(algorithms::gaussian_elimination::error::not_square);
+    }
+}
+
+
+/*
+ * description:
+ *      Finds inverse of matrix m containing fractions*/
+template <std::integral T, typename LP>
+auto inverse(ranges::matrix_view<ranges::fraction<T>, LP> m) -> std::expected<ranges::matrix_view<ranges::fraction<T>, LP>, algorithms::gaussian_elimination::error>{
```

Comment:
use gaussian elimination to implement inverse (do not duplicate code) - as during our classes extend matrix A to matrix [A | Id] and perform the algo. It is up to you how do you technically realize this extension [A | Id] ( e.g. you can create one big matrix which is probably the simples solution)